boa/src/syntax/ast/node/switch.rs

```diff
@@ -0,0 +1,85 @@
+//! Switch node.
+//!
+use super::{join_nodes, Node};
+use gc::{Finalize, Trace};
+use std::fmt;
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Clone, Debug, Trace, Finalize, PartialEq)]
+pub struct Switch {
+    val: Box<Node>,
+    cases: Box<[(Node, Box<[Node]>)]>,
+    default: Option<Box<Node>>,
+}
+
+/// The `switch` statement evaluates an expression, matching the expression's value to a case
+/// clause, and executes statements associated with that case, as well as statements in cases
+/// that follow the matching case.
+///
+/// A `switch` statement first evaluates its expression. It then looks for the first case
+/// clause whose expression evaluates to the same value as the result of the input expression
+/// (using the strict comparison, `===`) and transfers control to that clause, executing the
+/// associated statements. (If multiple cases match the provided value, the first case that
+/// matches is selected, even if the cases are not equal to each other.)
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///  - [MDN documentation][mdn]
+///
+/// [spec]: https://tc39.es/ecma262/#prod-SwitchStatement
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch
+impl Switch {
+    pub fn val(&self) -> &Node {
+        &self.val
+    }
+
+    pub fn cases(&self) -> &[(Node, Box<[Node]>)] {
+        &self.cases
+    }
+
+    pub fn default(&self) -> &Option<Box<Node>> {
+        &self.default
+    }
+
+    pub fn new(
+        val: Box<Node>,
+        cases: Box<[(Node, Box<[Node]>)]>,
+        default: Option<Box<Node>>,
```

Comment:
We should do this more generic, like we do with others, so that we can just pas something that's `Into<Node>` and then we box it inside this function.

Reply:
I tried like this,
`/// Creates a `Switch` AST node.
    pub fn new<V, C, D>(val: V, cases: C, default: D) -> Self
    where
        V: Into<Node>,
        C: Into<Box<[(V, Box<[V]>)]>>,
        D: Into<Option<V>>,
    {
        // expr: expr.into().map(E::into).map(Box::new),
        Self {
            val: Box::new(val.into()),
            cases: cases.into(),
            default: default.into().map(V::into).map(Box::new),
        }
    }`

but I'm struggling to think how to convert cases correctly.

I tried an as_ref, then map but not sure how to go from the map back into the box, do you have any suggestions?

Reply:
I would do it like this:
```rust
/// Creates a `Switch` AST node.
pub fn new<V, C, D>(val: V, cases: C, default: Option<V>) -> Self
where
    V: Into<Node>,
    C: Into<Box<[Case]>>,
{
    // expr: expr.into().map(E::into).map(Box::new),
    Self {
        val: Box::new(val.into()),
        cases: cases.into(),
        default: default.map(V::into).map(Box::new),
     }
}
```

Then, you need a `Case` struct like this one:
```rust
struct Case {
	condition: Node,
	statements: StatementList,
}
```

Reply:
Fixed as part of commit ea45b26 (to be merged), found that there was already a Case struct created for a similar purpose as part of SwitchStatement parsing so removed that struct and replaced it with the new Case struct.

Reply:
> Fixed as part of commit ea45b26 (to be merged), found that there was already a Case struct created for a similar purpose as part of SwitchStatement parsing so removed that struct and replaced it with the new Case struct.

We can have both, a `Case` parser and the produced `Case` Node, as we do with `Block` for example, if it makes sense.